[MRG+1] Add maximum streamline length property.

gzahl opened the pull request:
Hello,
I have added the option to specify the maximum streamline length. This is handy if one e.g. only needs a single (or a few) streamlines. It is especially handy if used together with "start_points".

This may be enhanced by forcing only forward (or backward) streamline integration. But that should probably be taken care of in another patch. Do you agree?

tacaswell commented:
Can you add a test for this?

Other than that seems perfectly reasonable and I agree doing the integration direction as a second PR is a good idea.   Smaller changes are easier to review.

gzahl commented:
I will try to set up a short test.

My first interface idea for the integration direction is to change maxlength optionally in a tuple:
maxlength=[0,2]  # this would be only forward integration.
maxlength=[-4,0] # only (longer) backward integration.
maxlength=2 # this would be translated to maxlength=[-2,2], which is the old default.

This would give the most flexibility. Or is it too complicated?

tacaswell commented:
The negatives would be confusing.  I think a tri-state `integration_direction in {'forward', 'reverse', 'both'}` and then the maximum_length is either all forward, all back or split evenly would be simpler and get _almost_ as much power.

gzahl commented:
Ok, i have added a test and copied the interface to axes and pyplot. I hope this was the right thing to do?
Tristate sounds good. Do you have an example, how it should be done in matplotlib? 
I think 'backward' is better than 'reverse', since it is the standard term when referring to integrations.

tacaswell commented:
I think just strings are good enough (just validate on the way in).

You will also need to commit the test image see http://matplotlib.org/devel/testing.html#writing-an-image-comparison-test

This should be a png only test I think.

gzahl commented:
Hi,
thanks for the all your help and quick replies! I now have added two more patches:
1. The baseline image for the maxlength test
2. A complete patch + test for the integration direction

Let me know, if anything has to be changed and i will try to do it tomorrow.

gzahl commented:
Ok, after ironing out my rookie mistakes, i have added two more commits.
Now only a single test fails, but it does not seem to be realted to my commits.
Is there anything i can do? Thanks!

tacaswell commented on lib/matplotlib/axes/_axes.py:
```diff
@@ -4465,21 +4465,25 @@ def stackplot(self, x, *args, **kwargs):
     def streamplot(self, x, y, u, v, density=1, linewidth=None, color=None,
                    cmap=None, norm=None, arrowsize=1, arrowstyle='-|>',
-                   minlength=0.1, transform=None, zorder=2, start_points=None):
+                   minlength=0.1, maxlength=4.0, transform=None, zorder=2,
```
All new args need to go at the end to maintain back-compatibility.  The reason is if people are (un wisely!) passing in everything as positional args.

When we can drop python2 _and_ can break API again most of these should be converted to kwarg only, but for now they need to go at the end.

tacaswell commented on lib/matplotlib/pyplot.py:
```diff
@@ -3267,8 +3267,9 @@ def step(x, y, *args, **kwargs):
 def streamplot(x, y, u, v, density=1, linewidth=None, color=None, cmap=None,
-               norm=None, arrowsize=1, arrowstyle='-|>', minlength=0.1,
-               transform=None, zorder=2, start_points=None, hold=None, data=None):
+               norm=None, arrowsize=1, arrowstyle='-|>', minlength=0.1, maxlength=4.0,
```
re-running boilerplate will take care of this.

tacaswell commented on lib/matplotlib/streamplot.py:
```diff
@@ -21,7 +21,8 @@
 def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
                cmap=None, norm=None, arrowsize=1, arrowstyle='-|>',
-               minlength=0.1, transform=None, zorder=2, start_points=None):
+               minlength=0.1, maxlength=4.0, transform=None, zorder=2,
```
following on a them, `maxlength` needs to go at the end here as well.

tacaswell commented on lib/matplotlib/streamplot.py:
```diff
@@ -91,6 +96,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     arrow_kw = dict(arrowstyle=arrowstyle, mutation_scale=10 * arrowsize)
 
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = "Integration direction " \
```
please use `()` for line continuation not `\`

tacaswell commented:
Left a few minor comments that need to be addressed.  I see what you are doing putting `minlength` and `maxlength` together, but we should not break API :disappointed: .

Could you also add an entry into https://github.com/matplotlib/matplotlib/tree/master/doc/users/whats_new so that this gets properly advertised in the next release?

gzahl commented:
Hi, Sorry for the delay. It seems like i don't get any Email notifications. I have hopefully addressed all the issues.
I will try to enable my notifications.

tacaswell commented on lib/matplotlib/streamplot.py:
```diff
@@ -52,9 +53,13 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     *minlength* : float
         Minimum length of streamline in axes coordinates.
+    *maxlength* : float
+        Maximum length of streamline in axes coordinates.
```
can you re-arrange the docstrings too?

gzahl commented:
The tests are failing after the merge with upstream. I have to track that down later.

afvincent commented:
Might the test failures be related to the fact that `FancyArrowPatch` (which is used for streamplots) was modified in parallel with your PR (#6504 that superseded #6054)?

tacaswell commented:
Sorry this got dropped!

Can you rebase instead of merging upstream into your branch? 

See http://matplotlib.org/devdocs/devel/gitwash/development_workflow.html?highlight=rebase

gzahl commented:
Thanks for the hint about a rebase rather than a merge.
But if i now revert the merge commit, i still have the merge/merge-revert in the history, which makes the rebase rather difficult (After conflicts with my changes, i get conflicts because of the merge commit).
Is there an easier way to do it? Sorry to bother you because of this. The easiest way would to make a new branch, and cherry-pick my changes. But i can not change the branch attached to this pull request.

WeatherGod commented:
If you do an interactive rebase, you can delete the commits that you don't
want in the branch and then force-push that up to github.

On Fri, Nov 4, 2016 at 7:04 AM, Manuel Jung <notifications@github.com>
wrote:

> Thanks for the hint about a rebase rather than a merge.
> But if i now revert the merge commit, i still have the merge/merge-revert
> in the history, which makes the rebase rather difficult (After conflicts
> with my changes, i get conflicts because of the merge commit).
> Is there an easier way to do it? Sorry to bother you because of this. The
> easiest way would to make a new branch, and cherry-pick my changes. But i
> can not change the branch attached to this pull request.
>
> —
> You are receiving this because you are subscribed to this thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/matplotlib/matplotlib/pull/6062#issuecomment-258403866>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AARy-AWfEMOjbTmqfAsmDjVSw-XVJLTsks5q6xEsgaJpZM4HjPdA>
> .
>

gzahl commented:
Ok, thanks for the hint about the interactive rebase. The rebase is done, but the test (of course) still fail.
I have looked into it for a bit, but i don't think this is (mainly) because of the #6504. The starting/end points seem to be slightly different. I will look into this asap.

afvincent commented:
If the diff images do not show any issue with the arrows (especially their size), then I think you are right about #6504 not being related to the test failures.

gzahl commented:
Looking at the history of streamplot.py and its test file, i make the following observations:
1. The test cases have been renamed bd63121d07cee1719248432ecf3aa61b2efb204f. I will do the same with mine, to comply to the others.
2. My observation about the different starting points seem to be related to commit 71b9ae8bd8b99843d67a635dde89e858d6f7d97b. Following this a new test case has been submitted 20959b3a6e50e255a8f5d066d286941dd3681715, which uses the starting_points option, which is also used by my tests. I therefore conclude, that this fix results in the changes of my test and will resubmit the expected image (Since it otherwise seems to be fine).

QuLogic commented:
Can you rebase and remove the extra copies of the test images?

Do an interactive rebase, move the update commits right below the original addition of the files, and change them to 'fixup'.

QuLogic commented on doc/users/whats_new/streamplot_set_maximum_length.rst:
```diff
@@ -0,0 +1,4 @@
+Maximum streamline length and integration direction can now be specified
+------------------------------------------------------------------------
+
+This allows to follow the vector field for a longer time and can enhance the visibilty of the flow pattern in some use cases.
```
visibilty -> visibility

Also, wrap at 80.

QuLogic commented on lib/matplotlib/pyplot.py:
```diff
@@ -2855,7 +2855,7 @@ def hexbin(x, y, C=None, gridsize=100, bins=None, xscale='linear',
 # changes will be lost
 @_autogen_docstring(Axes.hist)
-def hist(x, bins=10, range=None, normed=False, weights=None, cumulative=False,
+def hist(x, bins=None, range=None, normed=False, weights=None, cumulative=False,
```
Unrelated change?

gzahl replied:
Yes, but this is generated by boilerplate.py. Maybe it was changed in another commit, but boilerplate.py was not run? I rerun boilerplate.py in the 9634c0e.

Kojoley replied:
I think it is because `pyplot` was not updated with f1b89b209e9898e7f87ba3cb4f41b62aab6c82ed change

gzahl replied:
That does seem right. Should this be fixed first, in another issue?

tacaswell replied:
It's fine to update this here.  We should be better about not letting pyplot get 'stale', but it is what it is.

gzahl replied:
Okay. I guess this request can therefore be closed.
I rebase again yesterday. Again there are unrelated changes from boilerplate.py (4775d7e
).

QuLogic commented on lib/matplotlib/streamplot.py:
```diff
@@ -58,6 +59,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
         any number
+    *maxlength* : float
+        Maximum length of streamline in axes coordinates.
+    *integration_direction* : ['foward','backward','both']
```
foward -> forward

+ space after commas

QuLogic commented on lib/matplotlib/streamplot.py:
```diff
@@ -95,6 +100,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
 
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = ("Integration direction '%s' not recognised." %
+          integration_direction)
```
Misaligned.

QuLogic commented on lib/matplotlib/streamplot.py:
```diff
@@ -95,6 +100,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = ("Integration direction '%s' not recognised." %
+          integration_direction)
+        errstr += "Expected 'both', 'forward' or 'backward'."
```
Why not place it all in one string from the onset?

QuLogic commented on lib/matplotlib/streamplot.py:
```diff
@@ -435,17 +450,27 @@ def integrate(x0, y0):
+            dmap.reset_start_point(x0, y0)
+            s, xt, yt = _integrate_rk12(x0, y0, dmap, forward_time, maxlength)
+            if len(x_traj) > 0:
+              xt = xt[1:]
```
Indent should be a multiple of 4.

QuLogic commented on lib/matplotlib/tests/test_streamplot.py:
```diff
@@ -16,6 +16,14 @@ def velocity_field():
+def swirl_velocity_field():
+    x = np.linspace(-3.,3.,100)
+    y = np.linspace(-3.,3.,100)
+    X,Y = np.meshgrid(x,y)
```
Add space after commas in above 3 lines.

QuLogic commented on lib/matplotlib/tests/test_streamplot.py:
```diff
@@ -57,6 +65,23 @@ def test_masks_and_nans():
+def test_maxlength():
+    x, y, U, V = swirl_velocity_field()
+    plt.streamplot(x, y, U, V, maxlength=10., start_points=[[0.,1.5]],
+                   linewidth=2,density=2)
```
Add space after commas in above two lines.

QuLogic commented on lib/matplotlib/tests/test_streamplot.py:
```diff
@@ -57,6 +65,23 @@ def test_masks_and_nans():
+def test_direction():
+    x, y, U, V = swirl_velocity_field()
+    plt.streamplot(x, y, U, V, integration_direction='backward',
+                   maxlength=1.5, start_points=[[1.5,0.]],
```
Space after comma.

gzahl commented:
I have done the requested revisions.
I don't know what to do about the unrelated change in pyplot.py. It was autogenerated by boilerplate.py (see comment above). Thanks!

Kojoley commented on lib/matplotlib/streamplot.py:
```diff
@@ -95,6 +100,15 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
     arrow_kw = dict(arrowstyle=arrowstyle, mutation_scale=10 * arrowsize)
 
+    if integration_direction not in ['both', 'forward', 'backward']:
+        errstr = ("Integration direction '%s' not recognised."
```
You need a space after the dot.

gzahl replied:
Thanks, i fixed it (only line 104, right?)

gzahl commented:
@QuLogic Github still shows me a requested change, but i think i did all revisions? Otherwise please leave me a hint what is missing. Thanks!

NelleV commented on lib/matplotlib/streamplot.py:
```diff
@@ -58,6 +59,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
         In data coordinates, the same as the ``x`` and ``y`` arrays.
     *zorder* : int
         any number
+    *maxlength* : float
```
Can you please remove the "*". They are not helpful for readability.

gzahl replied:
Of course i could, but all parameters in this function use the "*". It seems inconsistent if i only remove them for the new "maxlength" parameter?

NelleV commented on lib/matplotlib/streamplot.py:
```diff
@@ -58,6 +59,10 @@ def streamplot(axes, x, y, u, v, density=1, linewidth=None, color=None,
         any number
+    *maxlength* : float
+        Maximum length of streamline in axes coordinates.
+    *integration_direction* : ['forward', 'backward', 'both']
```
Can you specify the default in the docstring?

QuLogic replied:
Oh, according to numpydoc, it should be braces `{}` instead of brackets `[]`, then the first value is the default.

NelleV commented on lib/matplotlib/streamplot.py:
```diff
@@ -401,7 +416,7 @@ class TerminateTrajectory(Exception):
 #========================
 
-def get_integrator(u, v, dmap, minlength):
+def get_integrator(u, v, dmap, minlength, maxlength, integration_direction):
```
I don't think this needs to be public. Can you make this function private?

gzahl replied:
Ok, if it was public, i should leave it like that?

NelleV commented:
Hi @gzahl 
It looks good overall, but do you mind making that new function private? Else, it adds this to the list of elements that need to be backward compatible :)
The "*" removal is just bonus points: we can merge this PR without that nitpick being fixed.

QuLogic commented:
Unfortunately, that function is not new (WRT this PR, at least.)

NelleV commented:
Indeed, it is not… (I am clearly not caffeinated enough).

tacaswell commented:
Glad this finally got merged!

@gzahl Thanks for your contribution and sorry the review took so long!

gzahl commented:
Thanks @tacaswell and all the others for your patience!
